### PR TITLE
[orchagent]: admin-disable port before setPortSerdesAttribute()

### PR DESCRIPTION
**Pull request by amnsinghal**
**What I did**
Admin-disable port before applying media-based NPU serdes attributes from [media_settings.json](https://github.com/sonic-net/SONiC/blob/master/doc/media-settings/Media-based-Port-settings.md).

**Why I did it**
This fix is needed along with xcvrd changes [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377).
Maintain deterministic behavior of [interface bring-up](https://github.com/sonic-net/SONiC/blob/master/doc/sfp-cmis/Interface-Link-bring-up-sequence.md), by toggling host_tx_ready flag, which will trigger CMIS reinit for the module once NPU serdes params have been applied.

**How I verified it**
Validated media_settings being notified and applied on Cisco 8111 with subject changes combined with diffs from [#360](https://github.com/sonic-net/sonic-platform-daemons/pull/360), [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377) and [#15453](https://github.com/sonic-net/sonic-buildimage/pull/15453).
Will update final results once [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377) is frozen.

**Details if related**
Proposal for xcvrd changes: [#356](https://github.com/sonic-net/sonic-platform-daemons/issues/356)

### CODE REVIEW

**Comment by amnsinghal** [comment]
@prgeor @mihirpat1 for your perusal.

**Review by eddyk-nvidia — APPROVED** [review]
Looks OK to me

**Review by Junchao-Mellanox — APPROVED** [review]

**Review by dgsudharsan — APPROVED** [review]

**Comment by prgeor** [comment]
@amnsinghal waiting for your UT test: Try this change in master without any other in-flight changes from Mihir's

**Comment by prgeor** [comment]
@amnsinghal @mihirpat1 we are seeing Xcvrd is missing the event host_tx_ready=False. Xcvrd is busy doing something else while OA sets the host_tx_ready = False and when finally OA sets the host_tx_ready=True, Xcvrd do get the notification but the filtering logic filters this event. Need to see a better way to notify change of media setting to Xcvrd.

**Comment by amnsinghal** [comment]
> @amnsinghal waiting for your UT test: Try this change in master without any other in-flight changes from Mihir's

@prgeor Here is a summary of UT results without Mihir's diff:
1. **Reboot, config reload**: All dockers/processes start clean, xcvrd starts CMIS FSM, notifies NPU media_settings to OA/syncd/SAI-SDK, link comes up as expected.
2. **xcvrd crash, xcvrd graceful restart**: xcvrd renotifies media_settings, causing link-flap, this should be taken care by [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377).
3. **syncd process crash, syncd docker restart**: xcvrd did not re-notify media_settings, nor did module DP reinit, this should be taken care by [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377). Link went down (due to NPU reset) and came up with default NPU params.
4. **syncd process graceful restart**: xcvrd did not re-notify media_settings, nor did module DP reinit, this should be taken care by [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377). Link went down (due to NPU reset) and came up with media-based NPU params from ASIC-DB.
5. **OA process crash**: xcvrd did not re-notify media_settings, nor did module DP reinit, this should be taken care by [#377](https://github.com/sonic-net/sonic-platform-daemons/pull/377). Link went down (due to NPU reset) and came up with default NPU params.

**Comment by amnsinghal** [comment]
@prgeor As discussed today, can you pls approve and merge this fix, if no further comments. Thanks.

File: orchagent/portsorch.cpp [hunk 1]
```diff
@@ -3996,17 +3996,34 @@ void PortsOrch::doPortTask(Consumer &consumer)
...
+                                    continue;
+                                }
+
+                                p.m_admin_state_up = false;
+                                m_portList[p.m_alias] = p;
+                        }
+
+                        if (setPortSerdesAttribute(p.m_port_id, gSwitchId, serdes_attr))
+                        {
+                            SWSS_LOG_NOTICE("Set port %s preemphasis is success", p.m_alias.c_str());
```
Comment (prgeor) [inline orchagent/portsorch.cpp:4017]:
```suggestion
                            SWSS_LOG_NOTICE("Set port %s SI settings is successful", p.m_alias.c_str());
```

File: orchagent/portsorch.cpp [hunk 2]
```diff
@@ -3996,17 +3996,34 @@ void PortsOrch::doPortTask(Consumer &consumer)
...
+                            p.m_preemphasis = serdes_attr;
+                            m_portList[p.m_alias] = p;
+                        }
+                        else
+                        {
+                            SWSS_LOG_ERROR("Failed to set port %s pre-emphasis", p.m_alias.c_str());
```
Comment (prgeor) [inline orchagent/portsorch.cpp:4023]:
```suggestion
                            SWSS_LOG_ERROR("Failed to set port %s SI settings", p.m_alias.c_str());
```

**Review by prgeor — APPROVED** [review]
@amnsinghal approved with minor comments

**Comment by prgeor** [comment]
@prsunny please help approve and merge.

**Comment by prgeor** [comment]
@amnsinghal re-submitted the test pipeline

**Comment by amnsinghal** [comment]
> @amnsinghal re-submitted the test pipeline

Thanks @prgeor

**Comment by amnsinghal** [comment]
@prgeor @prsunny Can you pls rerun Azure Pipeline, prev rerun failed in early build stage, apparently due to env/pipeline issue: https://dev.azure.com/mssonic/build/_build/results?buildId=323644&view=results. This changeset had passed all sanities yesterday. Thanks.

**Comment by lguohan** [comment]
test are failing, also new code needs to add new unit test code.

**Review by lguohan — CHANGES_REQUESTED** [review]
test are failing, also new code needs to add new unit test code.

**Comment by lguohan** [comment]
new code path must be protected by new unit test code, adding new logics without new code unit test is not a good practice. @prsunny

**Comment by amnsinghal** [comment]
> test are failing, also new code needs to add new unit test code.

@lguohan All tests had passed  2 days back before [Update syslog in orchagent/portsorch.cpp ](https://github.com/sonic-net/sonic-swss/pull/2831/commits/a5b94f3850f1c08ab42c9814eec19ab3768c1605), which was just a syslog rephrase, no functional change.

**Comment by amnsinghal** [comment]
> new code path must be protected by new unit test code, adding new logics without new code unit test is not a good practice. @prsunny

@lguohan This is existing code flow with only addition of admin-disable before applying NPU serdes attributes.

**Comment by prsunny** [comment]
@amnsinghal , can you please add unit test for coverage of this change?

**Comment by amnsinghal** [comment]
> @amnsinghal , can you please add unit test for coverage of this change?

@prsunny @lguohan Added UT in mock_tests/portsorch_ut.cpp, kindly review and approve.

**Review by prsunny — APPROVED** [review]

**Comment by amnsinghal** [comment]
Thanks @prgeor @prsunny @lguohan!
Can you pls request cherry-pick to 202205 and/or 202305.

**Comment by mihirpat1** [comment]
@StormLiangMS - Can you please help with cherry-pick to 202305?
ADO - 24988722

**Comment by StormLiangMS** [comment]
@amnsinghal  there is conflict, could you help to file separate PR for 202305?

**Pull request merged**